src/device/router_device_worker.erl

```diff
@@ -733,18 +732,45 @@ handle_cast(
                     end,
                 {devaddr, DevAddr0} = blockchain_helium_packet_v1:routing_info(Packet0),
                 DevAddr1 = lorawan_utils:reverse(<<DevAddr0:32/integer-unsigned-big>>),
+                Keys = router_device:keys(Device0),
                 DeviceUpdates = [
                     {keys,
                         lists:filter(
                             fun({NwkSKey, _}) -> NwkSKey == UsedNwkSKey end,
-                            router_device:keys(Device0)
+                            Keys
                         )},
                     {dev_nonces, DevNonces},
                     {devaddrs, [DevAddr1]},
                     {fcnt, PacketFCnt}
                 ],
                 D1 = router_device:update(DeviceUpdates, Device0),
                 ok = save_device(DB, CF, D1),
+
+                ToAdd = [{add, DevAddr0, UsedNwkSKey}],
+                %% We have to usort just in case DevAddr assigned is the same
+                ToRemove0 = lists:usort(
+                    lists:foldl(
+                        fun({NwkSKey, _}, Acc0) ->
+                            lists:foldl(
+                                fun(DevAddr, Acc1) ->
+                                    <<DevAddrInt:32/integer-unsigned-big>> = lorawan_utils:reverse(
+                                        DevAddr
+                                    ),
+                                    [{remove, DevAddrInt, NwkSKey} | Acc1]
+                                end,
+                                Acc0,
+                                router_device:devaddrs(Device0)
+                            )
+                        end,
+                        [],
+                        Keys
+                    )
+                ),
+                %% Making sure that the pair that was added is not getting removed (just in case DevAddr assigned is the same)
+                ToRemove1 = ToRemove0 -- [{remove, DevAddr0, UsedNwkSKey}],
```

Comment:
Making sure I understand this comment. 
Wouldn't this only be a problem if both DevAddr and NwkSKey are the same?

Should we flip around adding and removing just in case?
```erlang
ToRemove1 ++ ToAdd
```

Reply:
I don't know if the order thing matters anymore, on the other side the proplist is turned into a map.